### PR TITLE
JAMES-1951 lazy loading msn uid mapper

### PR DESCRIPTION
**Pull request by quynhn**

### CODE REVIEW

**Comment by quynhn** [comment]
test this please

**Comment by chibenwa** [comment]
```
[Building Java 8 james-1951_lazy_loading_msnuidmapper-8934bb61-463a-47bf-b224-33fb70de3da5] [INFO] Apache James MPT Imap Mailbox - JPA ................ FAILURE [ 18.916 s]

[Building Java 8 james-1951_lazy_loading_msnuidmapper-8934bb61-463a-47bf-b224-33fb70de3da5] testConcurrentExistsResponseUS(org.apache.james.mpt.imapmailbox.suite.ConcurrentSessions)  Time elapsed: 0.041 sec  <<< ERROR!
[Building Java 8 james-1951_lazy_loading_msnuidmapper-8934bb61-463a-47bf-b224-33fb70de3da5] org.apache.james.mpt.protocol.ProtocolSession$InvalidServerResponseException: 
[Building Java 8 james-1951_lazy_loading_msnuidmapper-8934bb61-463a-47bf-b224-33fb70de3da5] 
[Building Java 8 james-1951_lazy_loading_msnuidmapper-8934bb61-463a-47bf-b224-33fb70de3da5] Location: /org/apache/james/imap/scripts/ConcurrentExistsResponse.test:103
[Building Java 8 james-1951_lazy_loading_msnuidmapper-8934bb61-463a-47bf-b224-33fb70de3da5] LastClientMsg: 4b CREATE another
[Building Java 8 james-1951_lazy_loading_msnuidmapper-8934bb61-463a-47bf-b224-33fb70de3da5] Expected: '\* 1 FETCH \(FLAGS \(\\Flagged \\Recent\)\)'
[Building Java 8 james-1951_lazy_loading_msnuidmapper-8934bb61-463a-47bf-b224-33fb70de3da5] Actual   : '* 1 FETCH (FLAGS (\Recent))'
```

Looks related to your PR. Using *test this please* will not help...

**Comment by chibenwa** [comment]
Oups, triggered a new build :-(

File: protocols/imap/src/main/java/org/apache/james/imap/processor/base/SelectedMailboxImpl.java [hunk 1]
```diff
@@ -32,6 +32,9 @@
 import javax.mail.Flags;
 import javax.mail.Flags.Flag;
 
+import com.google.common.collect.DiscreteDomain;
```
Comment (strokyl) [inline protocols/imap/src/main/java/org/apache/james/imap/processor/base/SelectedMailboxImpl.java]: This commit is weird it look like it's partial and should be squashed somewhere

File: protocols/imap/src/main/java/org/apache/james/imap/processor/base/SelectedMailboxImpl.java [hunk 2]
```diff
@@ -99,13 +92,25 @@ public SelectedMailboxImpl(MailboxManager mailboxManager, ImapSession session, M
...
-            while(messages.hasNext()) {
-                MessageResult mr = messages.next();
-                add(mr.getUid());
-            }
-        }
+        uidMsnMapper = getUidMsnMapper(mailboxSession, messageManager);
+    }
+    
+    private UidMsnMapper getUidMsnMapper(MailboxSession mailboxSession , MessageManager messageManager) {
+                UidMsnMapper uidMsnMapper = new UidMsnMapper();
```
Comment (strokyl) [inline protocols/imap/src/main/java/org/apache/james/imap/processor/base/SelectedMailboxImpl.java]: Indent is weird

File: protocols/imap/src/main/java/org/apache/james/imap/processor/base/UidMsnMapper.java [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    private int getLastMsn() {
+        return getNumMessage();
+    }
+
+    //TODO: add unit test for that
+    public synchronized void remove(MessageUid uid) {
+        int msn = getMsn(uid).get();
+        msnToUid.remove(msn);
+
+        for (int aMsn = msn + 1; aMsn <= getNumMessage() + 1; aMsn++) {
```
Comment (strokyl) [inline protocols/imap/src/main/java/org/apache/james/imap/processor/base/UidMsnMapper.java]: `aMsn < getNumMessage()` would be better

File: protocols/imap/src/test/java/org/apache/james/imap/processor/base/UidMsnMapperTest.java [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+
+    @Test
+    public void getUidShouldReturnEmptyIfNoMessageWithTheGivenMessageNumber() {
+        assertThat(
+            testee.getUid(0).isPresent()
+        ).isFalse();
+    }
+
+    @Test
+    public void getUidShouldTheCorrespondingUidIfItExist() {
```
Comment (strokyl) [inline protocols/imap/src/test/java/org/apache/james/imap/processor/base/UidMsnMapperTest.java]: s/getUidShouldTheCorrespondingUidIfItExist/getgetUidShouldReturnTheCorrespondingUidIfItExist

File: protocols/imap/src/test/java/org/apache/james/imap/processor/base/UidMsnMapperTest.java [hunk 5]
```diff
@@ -156,146 +161,56 @@ public void addUidShouldNotOverridePreviousMapping() {
...
+        assertThat(testee.getInternals())
+            .isEqualTo(ImmutableBiMap.of(
+                1, messageUid2,
+                2, messageUid3,
+                3, messageUid4));
     }
 
     @Test
-    public void removeShouldKeepMessageNumberContiguousWhenDeletingTheLastOne() {
+    public void removeShouldKeepAValidMappingWhenDeletingMiddle() {
```
Comment (strokyl) [inline protocols/imap/src/test/java/org/apache/james/imap/processor/base/UidMsnMapperTest.java]: removeShouldKeepAValidMappingWhenDeletingEnd

File: protocols/imap/src/test/java/org/apache/james/imap/processor/base/UidMsnMapperTest.java [hunk 6]
```diff
@@ -156,146 +161,56 @@ public void addUidShouldNotOverridePreviousMapping() {
...
+        assertThat(testee.getInternals())
+            .isEqualTo(ImmutableBiMap.of(
+                1, messageUid1,
+                2, messageUid2,
+                3, messageUid3));
     }
 
     @Test
-    public void removeShouldKeepMessageNumberContiguousWhenDeletingOneInTheMiddle() {
+    public void removeShouldKeepAValidMappingWhenDeletingEnd() {
```
Comment (strokyl) [inline protocols/imap/src/test/java/org/apache/james/imap/processor/base/UidMsnMapperTest.java]: removeShouldKeepAValidMappingWhenDeletingMiddle

File: backends-common/jpa/src/test/java/org/apache/james/backends/jpa/JpaTestCluster.java [hunk 7]
```diff
@@ -74,9 +76,10 @@ public EntityManagerFactory getEntityManagerFactory() {
     public void clear(String... tables) {
         EntityManager entityManager = entityManagerFactory.createEntityManager();
         entityManager.getTransaction().begin();
-        for(String tableName: tables) {
+        entityManager.createNativeQuery("DROP ALL OBJECTS").executeUpdate();
+        /*for(String tableName: tables) {
```
Comment (strokyl) [inline backends-common/jpa/src/test/java/org/apache/james/backends/jpa/JpaTestCluster.java]: Comment should be removed

File: mailbox/jpa/src/test/java/org/apache/james/mailbox/jpa/JPAMailboxFixture.java [hunk 8]
```diff
@@ -35,7 +32,8 @@
         JPAProperty.class,
         JPAUserFlag.class,
         JPAMailboxAnnotation.class,
-        JPASubscription.class};
+        JPASubscription.class,
+        JPAMailboxAnnotationId.class};
```
Comment (strokyl) [inline mailbox/jpa/src/test/java/org/apache/james/mailbox/jpa/JPAMailboxFixture.java]: Then why don't you add MailboxIdUidKey.class if you added JPAMailboxAnnotationId.class ?

File: mailbox/jpa/src/test/java/org/apache/james/mailbox/jpa/mail/JPAMapperProvider.java [hunk 9]
```diff
@@ -47,16 +49,27 @@
...
+    private JpaTestCluster persistence;
+    private Exception e;
+
+    private JpaTestCluster createPersistence() {
+        //Preconditions.checkState(persistence == null);
+        if (persistence != null) {
+            return persistence;
+        }
+        persistence = JpaTestCluster.create(JPAMailboxFixture.MAILBOX_PERSISTANCE_CLASSES);
+        e = new Exception();
```
Comment (strokyl) [inline mailbox/jpa/src/test/java/org/apache/james/mailbox/jpa/mail/JPAMapperProvider.java]: WHY ?

Comment (strokyl) [inline mailbox/jpa/src/test/java/org/apache/james/mailbox/jpa/mail/JPAMapperProvider.java]: Look like you use it for debugging but you forgot to remove it no ?

File: dockerfiles/run/guice/jpa/destination/conf/META-INF/persistence.xml [hunk 10]
```diff
@@ -34,6 +34,8 @@
         <class>org.apache.james.domainlist.jpa.model.JPADomain</class>
         <class>org.apache.james.user.jpa.model.JPAUser</class>
         <class>org.apache.james.rrt.jpa.model.JPARecipientRewrite</class>
+        <class>org.apache.james.mailbox.jpa.mail.model.JPAMailboxAnnotation</class>
+        <class>org.apache.james.mailbox.jpa.mail.model.JPAMailboxAnnotationId</class>
```
Comment (strokyl) [inline dockerfiles/run/guice/jpa/destination/conf/META-INF/persistence.xml]: Why not `MailboxIdUidKey.class`

File: dockerfiles/run/spring/destination/conf/META-INF/persistence.xml [hunk 11]
```diff
@@ -31,6 +31,7 @@
         <class>org.apache.james.mailbox.jpa.mail.model.openjpa.JPAMailboxMessage</class>
         <class>org.apache.james.mailbox.jpa.mail.model.JPAProperty</class>
         <class>org.apache.james.mailbox.jpa.mail.model.JPAMailboxAnnotation</class>
+        <class>org.apache.james.mailbox.jpa.mail.model.JPAMailboxAnnotationId</class>
```
Comment (strokyl) [inline dockerfiles/run/spring/destination/conf/META-INF/persistence.xml]: Why not `MailboxIdUidKey.class`

File: protocols/imap/src/test/java/org/apache/james/imap/processor/base/UidMsnMapperTest.java [hunk 12]
```diff
@@ -122,7 +122,7 @@ public void isEmptyShouldReturnTrueIfNoMapping() {
     }
 
     @Test
-    public void isEmptyShouldReturnFalseIfNoMapping() {
+    public void isEmptyShouldReturnFalseIfMapping() {
```
Comment (strokyl) [inline protocols/imap/src/test/java/org/apache/james/imap/processor/base/UidMsnMapperTest.java]: I don't understand why you rename the test, also I don't understand what the test really test..

**Comment by chibenwa** [comment]
Mainstream PR merged for a long time

**Pull request closed without merging**